utils_test.go

```diff
@@ -42,3 +49,203 @@ func TestValidStoragePathFormat(t *testing.T) {
 		assert.NilError(t, err)
 	}
 }
+
+type HomeRuntime struct {
+	Dir   string
+	Error error
+}
+type CommandFile struct {
+	Data  string
+	Error error
+}
+type Stat struct {
+	Path        string
+	Replacement string
+}
+
+type RootlessRuntimeDirSyscallsTest struct {
+	RootlessRuntimeDirSyscallsImplementation
+	HomeRuntime HomeRuntime
+	CommandFile CommandFile
+	Stats       []Stat
+	Result      string
+}
+
+func (s RootlessRuntimeDirSyscallsTest) homeDirGetRuntimeDir() (string, error) {
+	return s.HomeRuntime.Dir, s.HomeRuntime.Error
+}
+func (s RootlessRuntimeDirSyscallsTest) ioUtilReadFile(path string) ([]byte, error) {
+	if path == "/proc/1/comm" {
+		return []byte(s.CommandFile.Data), s.CommandFile.Error
+	}
+	return ioutil.ReadFile(path)
+}
+func (s RootlessRuntimeDirSyscallsTest) systemLstat(path string) (*system.StatT, error) {
+	for _, Stat := range s.Stats {
+		if path == Stat.Path {
+			path = Stat.Replacement
+		}
+	}
+	return system.Lstat(path)
+}
+func (s RootlessRuntimeDirSyscallsTest) osStat(path string) (os.FileInfo, error) {
+	for _, Stat := range s.Stats {
+		if path == Stat.Path {
+			path = Stat.Replacement
+		}
+	}
+	return os.Stat(path)
+}
+
+func TestRootlessRuntimeDir(t *testing.T) {
+	rootlessUID := math.MaxUint16
+
+	dirForOwnerOnly, err := ioutil.TempDir(os.TempDir(), "containers-*")
+	assert.NilError(t, err)
+	defer os.Remove(dirForOwnerOnly)
+	err = os.Chmod(dirForOwnerOnly, 0700)
+	assert.NilError(t, err)
+
+	dirForAll, err := ioutil.TempDir(os.TempDir(), "containers-*")
+	assert.NilError(t, err)
+	defer os.Remove(dirForAll)
+	err = os.Chmod(dirForAll, 0777)
+	assert.NilError(t, err)
+
+	runUserDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+	tmpDir := fmt.Sprintf("%s/containers-user-%d", os.TempDir(), rootlessUID)
+	homeRunDir := homedir.Get() + "/rundir"
+
+	datas := []RootlessRuntimeDirSyscallsTest{
+		{
+			HomeRuntime: HomeRuntime{Dir: homeRunDir},
+			Result:      homeRunDir,
+		},
+
+		// systemd
+		{
```

Comment:
Non-blocking: Nothing in here exercises the “temporary directory was missing and successfully created” case I think. (But that would require moving the tests from real `os.TempDir()`, see the race test comment.)

Reply:
updated